File: oracle-js/ggl_legal.js
export class LegalityError extends Error {
  constructor(code, message) {
    super(message);
    this.code = code;
  }
}

export function checkLegality(ast, grammarAbi) {
  const legalPattern = grammarAbi.legal_regex;
  if (legalPattern) {
    const re = new RegExp(legalPattern, "s");
    if (!re.test(ast.text || "")) {
      throw new LegalityError("E_LEGAL_REGEX_NO_MATCH", "GGL text failed legality regex");
    }
  }
}

File: oracle/ggl_legal.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict


@dataclass
class LegalityError(Exception):
    code: str
    msg: str
    line: int = 0
    col: int = 0


def check_legality(ast: Dict[str, Any], grammar_abi: Dict[str, Any]) -> None:
    """Apply legality checks based on grammar ABI hints."""
    legal_pattern = grammar_abi.get("legal_regex")
    if legal_pattern:
        import re

        text = ast.get("text", "")
        if not re.fullmatch(legal_pattern, text, re.DOTALL):
            raise LegalityError("E_LEGAL_REGEX_NO_MATCH", "GGL text failed legality regex")
